### PR TITLE
feat(conf): Partial data export/import 

### PR DESCRIPTION
**Pull request by superrxan**

### CODE REVIEW

File: internal/binder/function/binder.go [hunk 1]
```diff
@@ -63,6 +64,18 @@ func Function(name string) (api.Function, error) {
 	return nil, e.GetError()
 }
 
+func GetFunctionPlugin(name string) (plugin.EXTENSION_TYPE, string, string) {
```
Comment (ngjaying) [inline internal/binder/function/binder.go:67]: Is manager.GetPluginMeta() can return the message you need?

File: internal/processor/ruleset.go [hunk 2]
```diff
@@ -30,7 +30,7 @@ type RulesetProcessor struct {
 type Ruleset struct {
 	Streams map[string]string `json:"streams"`
 	Tables  map[string]string `json:"tables"`
-	Rules   map[string]string `json:"rules"`
+	Rules   map[string]string `json:"Rules"`
```
Comment (ngjaying) [inline internal/processor/ruleset.go:33]: why change the cap?

File: internal/binder/factory.go [hunk 3]
```diff
@@ -14,15 +14,28 @@
...
+)
 
 type SourceFactory interface {
 	Source(name string) (api.Source, error)
 	LookupSource(name string) (api.LookupSource, error)
+	// GetSourcePlugin use source type name (mqtt/video) to get source plugin installation information
+	// first return value is extension type, only native/portable plugin type have installation information
+	// second return value is the plugin name
+	// third is the plugin installation information
+	GetSourcePlugin(name string) (plugin.EXTENSION_TYPE, string, string)
```
Comment (ngjaying) [inline internal/binder/factory.go:29]: rename to SourcePluginInfo? To be consistent as the other two func names. The same for other factories

File: internal/plugin/portable/factory.go [hunk 4]
```diff
@@ -30,6 +30,17 @@ func (m *Manager) Source(name string) (api.Source, error) {
 	return runtime.NewPortableSource(name, meta), nil
 }
 
+func (m *Manager) GetSourcePlugin(name string) (plugin.EXTENSION_TYPE, string, string) {
```
Comment (ngjaying) [inline internal/plugin/portable/factory.go:33]: Is GetPluginMeta the same as this?

File: internal/plugin/native/manager.go [hunk 5]
```diff
@@ -585,6 +585,19 @@ func (rr *Manager) Source(name string) (api.Source, error) {
 	}
 }
 
+func (rr *Manager) GetSourcePlugin(name string) (plugin2.EXTENSION_TYPE, string, string) {
```
Comment (ngjaying) [inline internal/plugin/native/manager.go:588]: GetPluginInfo seems similar to this one

**Review by ngjaying — COMMENTED** [review]
some comments and suggestions

File: internal/plugin/native/manager.go (hunk 5 above)
Comment (superrxan) [reply internal/plugin/native/manager.go:588]: GetPluginInfo do not have the install infomation of the plugin
 GetSourcePlugin need the install infomation of the plugin

File: internal/plugin/portable/factory.go (hunk 4 above)
Comment (superrxan) [reply internal/plugin/portable/factory.go:33]: GetPluginMeta do not have the install infomation of the plugin
 GetSourcePlugin need the install infomation of the plugin

File: internal/binder/factory.go (hunk 3 above)
Comment (superrxan) [reply internal/binder/factory.go]: ack

**Review by ngjaying — CHANGES_REQUESTED** [review]

**Review by ngjaying — CHANGES_REQUESTED** [review]
Some comments and suggestions, please have a look

File: cmd/kuiper/main.go [hunk 6]
```diff
@@ -1054,7 +1065,7 @@ func main() {
 		{
 			Name:    "export",
 			Aliases: []string{"export"},
-			Usage:   "export ruleset | data $ruleset_file",
+			Usage:   "export ruleset | data $ruleset_file [ -r rules ]",
```
[the remark this answers is not in the document]
Comment (superrxan) [reply cmd/kuiper/main.go:1068]: line 1104 to 1125 handle this case

File: internal/plugin/native/manager.go [hunk 7]
```diff
@@ -846,6 +879,21 @@ func (rr *Manager) clearInstallFlag() {
...
+	}
+	err = rr.Register(plgType, sd)
+	if err != nil {
+		conf.Log.Errorf(`install native plugin %s error: %v`, key, err)
+		return err
+	}
+	return nil
+}
+
 func (rr *Manager) pluginInstallWhenReboot() {
```
[the remark this answers is not in the document]
Comment (superrxan) [reply internal/plugin/native/manager.go:897]: native plugin do not install when import from file, only save the install info into db.
when ekuiper reboot, willl read the plugin info from db and install.

File: internal/schema/registry.go [hunk 8]
```diff
@@ -299,6 +299,37 @@ func ImportSchema(schema map[string]string) error {
 	return schemaDb.Set(BOOT_INSTALL, BOOT_INSTALL)
 }
 
+// SchemaPartialImport compare the schema to be installed and the one in database
+// if not exist in database, install;
+// if exist, ignore
+func SchemaPartialImport(schemas map[string]string) map[string]string {
```
[the remark this answers is not in the document]
Comment (superrxan) [reply internal/schema/registry.go:305]: yes, already supported

File: internal/server/rule_migration.go [hunk 9]
```diff
@@ -0,0 +1,570 @@
...
+			}
+			return true
+		})
+
+		//Rules
+		de.rules = append(de.rules, rule.Id)
+	}
+
+	ruleGraph := rule.Graph
+	if ruleGraph != nil {
```
[the remark this answers is not in the document]
Comment (superrxan) [reply internal/server/rule_migration.go:153]: ack

**Review by ngjaying — APPROVED** [review]

**Pull request merged**